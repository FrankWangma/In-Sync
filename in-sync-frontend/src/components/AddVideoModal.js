import React, { useState } from "react";
import {
  Button,
  TextField,
  Typography,
  Grid,
  Dialog,
  DialogContent,
  DialogTitle,
  DialogActions
} from "@material-ui/core";
import styles from "./Modal.module.css";
import VideoList from './VideoList'
import axios from 'axios';

const AddVideoModal = ({ showModal, modalHandler, handleVideoChange }) => {
  const [url, setVideoURL] = useState("");
  const [search, setSearchInput] = useState("");
  const [videos, setVideos] = useState([]);
  const [error, setError] = useState(false);
  const [helperText, setHelperText] = useState("");
  const KEY = process.env.YOUTUBE_KEY || "API KEY NOT VALID LOCALLY";

  const youtube = axios.create({
    baseURL: 'https://www.googleapis.com/youtube/v3',
    params: {
      part: 'snippet',
      maxResults: 5,
      key: KEY,
    }
  })

  const checkYoutubeUrl = (input) => {
    // eslint-disable-next-line no-useless-escape
    const re = /^((?:https?:)?\/\/)?((?:www|m)\.)?((?:youtube\.com|youtu.be))(\/(?:[\w\-]+\?v=|embed\/|v\/)?)([\w\-]+)(\S+)?$/;
    return re.test(input);
  }

  const handleSearchSubmit = () => {
    youtube.get('/search', {
      params: {
        part: 'snippet',
        maxResults: 5,
        key: KEY,
        type: 'video',
        q: search
      }
    }).then((response) => {
      setVideos(response.data.items);
    })
  }

  const handleVideoSelect = (video) => {
    const url = `https://www.youtube.com/watch?v=${video.id.videoId}`;
    handleVideoChange(url);
    modalHandler(false)
  }

  const handleChangeKey = (e) => {
    if (keyPress(e) && url) {
      changeVideos();
    }
  }

  const handleSearchKey = (e) => {
    if (keyPress(e) && search) {
      handleSearchSubmit();
    }
  }

  const changeVideos = () => {
    if (checkYoutubeUrl(url)) {
      handleVideoChange(url);
      setError(false);
      setHelperText('');
      setVideoURL('');
      modalHandler(false);
    } else {
      setError(true);
      setHelperText('Please enter a valid YouTube URL');
    }
  }

  const keyPress = (e) => {
    if (e.keyCode === 13) {
      return true;
    }
    return false;
  }

  return (
    <Dialog
      open={showModal}
      onBackdropClick={() => {
        modalHandler(false);
        setError(false);
        setHelperText('');
        setVideoURL('');
      }}
      scroll='paper'
      fullWidth={true}
      maxWidth={'lg'}
    >
      <DialogTitle>Change the video</DialogTitle>
      <DialogContent className={styles.changeVideoModal} dividers={true}>
        <Grid container spacing={0}>
          <Grid item xs={1} />
          <Grid item xs={5}>
            <Typography>Video URL</Typography>
            <TextField
              error={error}
              helperText={helperText}
              className={styles.addVideoText}
              InputProps={{ disableUnderline: true }}
              margin="normal"
              name="videoURL"
              value={url}
              placeholder="https://www.youtube.com/watch?v=dQw4w9WgXcQ"
              onChange={(e) => { setVideoURL(e.target.value); }}
              onKeyDown={handleChangeKey}
            />
            <Typography>Search for a video</Typography>
            <TextField
              className={styles.addVideoText}
              InputProps={{ disableUnderline: true }}
              margin="normal"
              name="videoSearch"
              value={search}
              onChange={(e) => { setSearchInput(e.target.value); }}
              onKeyDown={handleSearchKey}
            />
          </Grid>
          <Grid item xs={1} />
          <Grid item xs={4}>
            <Typography variant="h6">
              Top 5 Videos Found
            </Typography>
            <VideoList handleVideoSelect={handleVideoSelect} videos={videos} />
          </Grid>
          <Grid item xs={1} />
        </Grid>
      </DialogContent>
      <DialogActions>
        <Button className={styles.cancelButton} onClick={() => { modalHandler(false); }}>
          Cancel
        </Button>
        <Button variant="contained" className={styles.changeButton} disabled={!url} onClick={changeVideos}>
          Change
        </Button>
        <Button variant="contained" disabled={!search} onClick={handleSearchSubmit}>
          Search
        </Button>
      </DialogActions>
    </Dialog>
  );
};

export default AddVideoModal;
