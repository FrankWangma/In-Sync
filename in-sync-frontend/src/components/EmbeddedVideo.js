import React, { useRef } from 'react';
import { findDOMNode } from 'react-dom';
import ReactPlayer from 'react-player';
import Grid from '@material-ui/core/Grid';
import screenfull from 'screenfull';
import Button from '@material-ui/core/Button';
import { Fullscreen, VolumeUp, PlayArrow, Pause } from '@material-ui/icons';
import Slider from '@material-ui/core/Slider';

const EmbeddedVideo = (props) => {

    const [playing, setPlaying] = React.useState(false);
    const [volume, setVolume] = React.useState(50);
    const [played, setPlayed] = React.useState(0);
    const [seeking, setSeeking] = React.useState(false);

    const ref = useRef();

    const handlePlay = () => {
        setPlaying(true);
        props.playVideo(this.state.played);
    }

    const handlePause = () => {
        setPlaying(false);
        props.pauseVideo(this.state.played);
    }

    const handlePlayPause = () => {
        setPlaying(!playing);
    }

    const handleSeekMouseDown = _e => {
        setSeeking(true);
    }

    const handleSeekChange = (_e, v) => {
        ref.current.seekTo(v * 2);
        setPlayed(v);
    }

    const handleSeekMouseUp = (e, v) => {
        setSeeking(false);
    }

    const handleVolumeChange = (_e, v) => {
        setVolume(v);
    }

    const handleProgress = state => {
        // We only want to update time slider if we are not currently seeking
        if (!seeking) {
            setPlayed(state.played);
        }
    }

    const handleClickFullscreen = () => {
        screenfull.request(findDOMNode(ref.current));
    }

    return (
        <div className="App EmbeddedVideo">
            <ReactPlayer
                url={props.url}
                ref={ref}
                playing={playing}
                volume={volume / 100}
                onPlay={handlePlay}
                onPause={handlePause}
                onEnded={handlePause}
                onProgress={handleProgress}
            />

            {/* Video Player Controls */}
            <Grid container spacing={0}>
                <Grid item xs={2}>
                    <Button variant="contained" onClick={handlePlayPause}>{playing ? <Pause /> : <PlayArrow />}</Button>
                </Grid>
                <Grid item xs={8}>
                    <Slider
                        value={played * 100}
                        onMouseDown={handleSeekMouseDown}
                        onChange={handleSeekChange}
                        onMouseUp={handleSeekMouseUp}
                        aria-labelledby="continuous-slider" />
                </Grid>
                <Grid item xs={2}>
                    <Button variant="contained" onClick={handleClickFullscreen}><Fullscreen /></Button>
                </Grid>
                <Grid item xs={2}>
                    <p className='volume'><VolumeUp /></p>
                </Grid>
                <Grid item xs={3}>
                    <Slider value={volume} onChange={handleVolumeChange} aria-labelledby="continuous-slider" />
                </Grid>
            </Grid>
        </div>
    );
}

export default EmbeddedVideo;
