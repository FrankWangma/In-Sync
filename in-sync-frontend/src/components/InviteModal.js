import React, { useRef } from "react";
import {
  Typography,
  Modal,
} from "@material-ui/core";
import styles from "./Modal.module.css";

const InviteModal = ({ showModal, modalHandler, roomId }) => {
  const link = window.location.host + "/joinRoom?id=" + roomId;
  const codeRef = useRef(null);
  const linkRef = useRef(null);

  // const copyCodeToClipboard = (e) => {
  //   codeRef.current.select();
  //   document.execCommand('copy');
  //   e.target.focus();
  // }

  // const copyLinkToClipboard = (e) => {
  //   linkRef.current.select();
  //   document.execCommand('copy');
  //   e.target.focus();
  // }

  return (
  <Modal open={showModal} onBackdropClick={() => { modalHandler(false); }}>
    <div className={styles.appModal}>
      <Typography className={styles.title}>
        To invite users to the room they can use the code:
      </Typography>
      <Typography ref={codeRef}>{roomId}</Typography>
      {/* {
        document.queryCommandSupported('copy') &&
          <Button onClick={copyCodeToClipboard}>Copy</Button>
      } */}
      <Typography className={styles.title}>
        Or use the link:
      </Typography>
      <Typography ref={linkRef}>{link}</Typography>
      {/* {
        document.queryCommandSupported('copy') &&
          <Button onClick={copyLinkToClipboard}>Copy</Button>
      } */}
    </div>
  </Modal>
  );
};

export default InviteModal;
