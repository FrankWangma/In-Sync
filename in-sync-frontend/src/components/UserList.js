import React from "react";
import styles from  "./UserList.module.css";

// Users would populate from backend
const users = [
  {
    name: "Ironman",
    uid: "user1",
    avatar: "https://data.cometchat.com/assets/images/avatars/ironman.png",
  },
  {
    name: "Spiderman",
    uid: "user2",
    avatar: "https://data.cometchat.com/assets/images/avatars/spiderman.png",
  },
];

const UserList = () => {
  const displayUser = (user) =>
    (<div className={styles.userInfo}>
      <img className={styles.userPic} src={user.avatar} alt="User avatar" />
      <p className={styles.userName}>{user.name}</p>
    </div>);

  return (
    <div className={styles.container}>
      {users.map((user, index) => <div key={index} >{displayUser(user)}</div>)}
    </div>
  );
};

export default UserList;
