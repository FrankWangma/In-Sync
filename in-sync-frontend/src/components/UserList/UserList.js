import React from "react";
import Star from "@material-ui/icons/Star";
import styles from "./UserList.module.css";

const avatar = "http://getdrawings.com/free-icon-bw/free-avatars-icons-6.png";
const UserList = ({ users }) => {
  const displayUser = (user) => (
      <div className={styles.userInfo}>
        <img className={styles.userPic} src={avatar} alt="User avatar" />
        {user === users.host ? <Star className={styles.hostStar}/> : <div />}
        <p className={styles.userName}>{user}</p>
      </div>
  );

  return (
    <div className={styles.container}>
      {[...new Set(users.viewers)].map((user, index) => <div key={index} >
      {displayUser(user)}
      </div>)}
    </div>
  );
};

export default UserList;
