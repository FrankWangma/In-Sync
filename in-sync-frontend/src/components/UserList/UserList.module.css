.userInfo {
    background-color: #253040;
    display: flex;
    flex-direction: row;
    height: 50px;
    margin: 10px;
    border-radius: 10px;
}

.userPic {
    width: 40px;
    height: 40px;
    float: left;
    margin: 5px 0 0 5px;
}

.userName {
    margin-left: 10px;
}

.hostStar {
    color: yellow;
    margin: 12px 0 0 10px;
}

.container {
    background-color: #F8F9FB;
    margin: 0 15px;
    width: calc(100% - 30px);
    max-height: 595px;
    min-height: 595px;
    overflow-y: auto;
    border-bottom-left-radius: 7px;
    border-bottom-right-radius: 7px;
}
