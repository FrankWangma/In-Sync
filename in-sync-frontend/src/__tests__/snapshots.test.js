import React from 'react';
import renderer from 'react-test-renderer';

import AddVideoModal from '../components/AddVideoModal';
import UserList from '../components/UserList';
import InviteModal from '../components/InviteModal';
import EmbeddedVideo from '../components/EmbeddedVideo';


it(' AddVideoModal renders correctly with no content', () => {
  const tree = renderer.create(<AddVideoModal />).toJSON();
  expect(tree).toMatchSnapshot();
});

it(' UserList renders correctly', () => {
  const tree = renderer.create(<UserList />).toJSON();
  expect(tree).toMatchSnapshot();
});

it(' InviteModal renders correctly', () => {
  const tree = renderer.create(<InviteModal />).toJSON();
  expect(tree).toMatchSnapshot();
});

it(' EmbeddedVideo renders correctly', () => {
    const tree = renderer.create(<EmbeddedVideo />).toJSON();
    expect(tree).toMatchSnapshot();
});