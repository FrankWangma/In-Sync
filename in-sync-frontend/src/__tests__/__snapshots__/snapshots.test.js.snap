// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` AddVideoModal renders correctly with no content 1`] = `null`;

exports[` EmbeddedVideo renders correctly 1`] = `
<div
  className="App EmbeddedVideo"
>
  <div
    style={
      Object {
        "height": "360px",
        "width": "640px",
      }
    }
  />
  <div
    className="MuiGrid-root MuiGrid-container"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
    >
      <button
        onClick={[Function]}
      >
        â–¶
      </button>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
    >
      <input
        className="seek"
        max={0.999999}
        min={0}
        onChange={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        step="any"
        type="range"
        value={0}
      />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
    >
      <button
        onClick={[Function]}
      >
        â›¶
      </button>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
    >
      <p
        className="volume"
      >
        ðŸ”Š
      </p>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
    >
      <input
        max={1}
        min={0}
        onChange={[Function]}
        step="any"
        type="range"
        value={0.5}
      />
    </div>
  </div>
</div>
`;

exports[` InviteModal renders correctly 1`] = `null`;

exports[` UserList renders correctly 1`] = `
<div
  className="container"
>
  <div>
    <div
      className="userInfo"
    >
      <img
        alt="User avatar"
        className="userPic"
        src="https://data.cometchat.com/assets/images/avatars/ironman.png"
      />
      <p
        className="userName"
      >
        Ironman
      </p>
    </div>
  </div>
  <div>
    <div
      className="userInfo"
    >
      <img
        alt="User avatar"
        className="userPic"
        src="https://data.cometchat.com/assets/images/avatars/spiderman.png"
      />
      <p
        className="userName"
      >
        Spiderman
      </p>
    </div>
  </div>
</div>
`;
